packages/SwingSet/test/virtualObjects/vat-vom-gc-bob.js

```diff
@@ -0,0 +1,59 @@
+/* global makeKind */
+import { E } from '@agoric/eventual-send';
+
+const things = [];
+
+export function buildRootObject(_vatPowers) {
+  function makeThingInstance(state) {
+    return {
+      init(label) {
+        state.label = label;
+      },
+      self: {
+        getLabel() {
+          return state.label;
+        },
+      },
+    };
+  }
+
+  const thingMaker = makeKind(makeThingInstance);
+  let nextThingNumber = 0;
+
+  return harden({
+    prepare() {
+      for (let i = 1; i <= 9; i += 1) {
+        things.push(thingMaker(`thing #${i}`));
+      }
+    },
+    getThing(forWhom) {
+      let thing;
+      do {
+        thing = things[nextThingNumber];
+        nextThingNumber += 1;
+      } while (!thing);
+
+      if (nextThingNumber === 2) {
+        thing.getLabel();
+        things[3].getLabel();
+        things[3] = null; // arbitrarily drop one before sending it, but don't drop the one we're sending
+      } else {
+        thing.getLabel();
+        things[nextThingNumber - 1] = null; // drop the one we're sending
+      }
+      E(forWhom).deliverThing(thing);
+      thing = null;
+    },
+    finish() {
+      while (nextThingNumber < 7) {
+        const deadThing = things[nextThingNumber];
+        if (deadThing) {
+          deadThing.getLabel();
+          things[nextThingNumber] = null;
+        }
+        nextThingNumber += 1;
+      }
+      console.log(`Bob finishing`);
```

Comment:
I find this test pretty confusing. If I read it correctly, by the time `finish()` finishes, the virtual objects originally associated with the Representatives in `things[]` are left in the following states
| number | exported | has Representative | in virtualized data | notes |
| ---- | ---- | ---- | ---- | ---- |
| 1 | exported+dropped | no | no | Rep dropped before kernel drops |
| 2 | exported+dropped | yes | no | |
| 3 | never | no | no | |
| 4 | exported+retained | no | no | |
| 5 | never | no | no | |
| 6 | never | no | no | |
| 7 | never | yes | no | no | |
| 8 | never | yes | no | no | |
| 9 | never | yes | no | no | |

The test is asserting:
* `v1.vs.vom.o+1/2` (that's thing2, right?) has a 0/0 refcount: good, it's kept alive by the Representative pillar and nothing else
* `../3` (thing3) has a 1/0 refcount .. wait, thing4 should have 1/0 (kept alive by the kernel-export pillar, but no Representative), am I off-by-one already? The `vobjID` are assigned upon object creation, not export, yeah?
* `../4` through `../7` are deleted: I'm off again, it looks like thing7 should be retained by a Representative
* `../8` and `../9` have data but no refcounts, which seems slightly weird, but I guess you're avoiding the DB expense for virtual objects that are retained by only a Representative and not by either a kernel-export or virtualized-data?

If we can enumerate the cases we care about, it might be easier to follow this test (and to enhance it in the future) to make them all explicit. Create a batch of Representatives, forget a few, export a few, store a few in virtualized data. Wait for the peer to drop some of the exported ones. Forget a few more (to test both orderings: Representative collected before kernel drops, kernel drops before Representative collected). Then let the test runner look at the DB and make sure everything matches the expectations.

Reply:
These cases will remain confusing until we have implemented all three legs of the tripod.  Remember that at this point we've only got 2, because using something in a property value pins it forever.

Reply:
Let me see if I can make this clearer (I'll turn this into a comment in the test if so):

First of all, nothing in this test exercises a case where any virtual object refers to any other virtual object, so the fourth column of your table is nugatory.  Also, I'm not sure what the batch of "no"s in the lower portion of the table in the *Notes* column even means.  With respect to any given case, we really only have two variables to fiddle with: whether a reference to the VO is exported, and whether a reference to the VO is retained in memory in the vat.

For the first variable, there are three possible states: not exported, exported and then retained externally, exported and then dropped externally.

For the second variable, there are two possible states: retained internally, and dropped internally.

That would seem like 6 cases, but that's not quite true, since a reference that is dropped internally before being exported can't ever be exported.  (Also if a VO has both an exported reference and an internal reference and both are dropped, we still have to worry about what order they're dropped in, but it's the other test that worries about the order of droppage issue).

| number | exported? | export dropped? | local dropped? | deleted?
| ---- | ---- | ---- | ---- | ---- |
| 1 | yes | yes | yes | yes (dropped both locally and by export)|
| 2 | yes | yes | no  | no (retained in `things`)|
| 3 | yes | no | yes | no (retained by export)|
| 4 | no | n/a | yes | yes (dropped before use) |
| 5,6,7 | no | n/a | yes | yes (dropped at end) |
| 8,9 | no | n/a | no | no (retained unused) |

Things 5, 6, and 7 are essentially the same as 4, but they were handled in a loop at the end, to make sure that that worked.
Things 8 and 9 are both the same.  They were never used and so were retained where they were originally stashed on creation.

Reply:
Ooookay, I think I see how that table might come out of the logic in `getThing`, but I don't think I'd be able to figure that out by reading the test (and obviously I've already gotten it wrong once). Thank's for adding the table to the test case, that will help me a lot.